tasks/model.go

```diff
@@ -197,42 +215,102 @@ func (tp *_Task__Prototype) New(r RetrievalTask, s StorageTask) Task {
 	return &t
 }
 
-func (t *_Task) Assign(worker string, status Status) {
-	t.WorkedBy = _String__Maybe{m: schema.Maybe_Value, v: &_String{worker}}
-	t.StartedAt = _Time__Maybe{m: schema.Maybe_Value, v: &_Time{x: time.Now().UnixNano()}}
-	t.Status = *status
+func (t *_Task) Assign(worker string, status Status) Task {
+	newTask := _Task{
+		UUID:                t.UUID,
+		Status:              *status,
+		WorkedBy:            _String__Maybe{m: schema.Maybe_Value, v: &_String{worker}},
+		Stage:               t.Stage,
+		CurrentStageDetails: t.CurrentStageDetails,
+		PastStageDetails:    t.PastStageDetails,
+		StartedAt:           _Time__Maybe{m: schema.Maybe_Value, v: &_Time{x: time.Now().UnixNano()}},
+		RetrievalTask:       t.RetrievalTask,
+		StorageTask:         t.StorageTask,
+	}
 
 	//todo: sign
+	return &newTask
 }
 
-func (t *_Task) Update(status Status, stage string, details StageDetails) error {
-	t.Status = *status
-	t.Stage = _String{stage}
+func (t *_Task) Update(status Status, stage string, details StageDetails) (Task, error) {
+	updatedTask := _Task{
+		UUID:          t.UUID,
+		Status:        *status,
+		WorkedBy:      t.WorkedBy,
+		Stage:         _String{stage},
+		StartedAt:     t.StartedAt,
+		RetrievalTask: t.RetrievalTask,
+		StorageTask:   t.StorageTask,
+	}
+
+	// On stage transitions, archive the current stage.
+	if stage != t.Stage.x && t.CurrentStageDetails.Exists() {
+		if !t.PastStageDetails.Exists() {
+			updatedTask.PastStageDetails = _List_StageDetails__Maybe{m: schema.Maybe_Value, v: &_List_StageDetails{x: []_StageDetails{*t.CurrentStageDetails.v}}}
+		} else {
+			updatedTask.PastStageDetails = _List_StageDetails__Maybe{m: schema.Maybe_Value, v: &_List_StageDetails{x: append(t.PastStageDetails.v.x, *t.CurrentStageDetails.v)}}
+		}
+	}
+
 	if details == nil {
-		t.CurrentStageDetails = _StageDetails__Maybe{m: schema.Maybe_Absent}
+		updatedTask.CurrentStageDetails = _StageDetails__Maybe{m: schema.Maybe_Absent}
 	} else {
-		t.CurrentStageDetails = _StageDetails__Maybe{m: schema.Maybe_Value, v: details}
+		updatedTask.CurrentStageDetails = _StageDetails__Maybe{m: schema.Maybe_Value, v: details}
 	}
 
 	//todo: sign
-	return nil
+	return &updatedTask, nil
 }
 
-func (t *_Task) UpdateTask(tsk UpdateTask) error {
-	t.Status = tsk.Status
-	if tsk.CurrentStageDetails.Exists() {
-		t.CurrentStageDetails = tsk.CurrentStageDetails
-	} else {
-		t.CurrentStageDetails = _StageDetails__Maybe{m: schema.Maybe_Absent}
+func (t *_Task) Finalize(ctx context.Context, s ipld.Storer) (FinishedTask, error) {
+	if t.Status != *Failed && t.Status != *Successful {
```

Comment:
I'm a bit uneasy that we always dereference these globals when using them. Maybe they shouldn't be pointers.

Reply:
i think this is a consequence of how ipld deals with exported types (the 'Status' vs '_Status' thing)